### PR TITLE
Prisma change over feature branch 23 01 29

### PR DESCRIPTION
**Pull request by ludtkemorgan**
# Pull Request Template

## Issue Overview

This PR addresses #issue

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

This pulls the following commits from main:
* 42cd7fe52 - (origin/main) fix: add divider width (#3841) <ColinBuyck>
* 99847f07c - feat: 422/updated account theming (#3791) <ColinBuyck>
* 3ea47eda1 - (main) fix: Paper Application not saving lat long (#3828) <Krzysztof Zięcina>
* 09612429e - fix: add submit to sms mfa flow (#3838) <Morgan Ludtke>
* e3305e413 - feat: 3761/direct to sign in (#3816)  <cade-exygy>
* c3a6b9c67 - fix: geocoding verification fix (#3832) <Morgan Ludtke>
* 27c13b6e6 - fix: add default value for map layer (#3829) <Morgan Ludtke>
* 6fe537422 - fix: add map layer geocoding checks (#3825) <Morgan Ludtke>
* aa8e3a647 - feat: geocoding preference set up fixed layers (#3808) <Krzysztof Zięcina>
* 66816871e - fix: seeds button layout fixes for Partners Application sidebar (#3789) <Jared White>
* af18a9bb8 - fix: add links to preference description (#3813) <Krzysztof Zięcina>
* 87288ef56 - fix: provides additional fields info text overflows (#3809) <Krzysztof Zięcina>

Also implements the remaining portions of geocoding

## How Can This Be Tested/Reviewed?

The main thing that needs to be tested is the geocoding functionality. This can be tested by reseeding with stage data and applying to the "Elm Village" property for both public and the partner site. It should add the geocoding verified value on the export.

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

**Comment by ColinBuyck** [comment]
Still reviewing but went to test the edit application to include geocoding preference data and noticed an issue with saving after editing a paper application now. Steps to reproduce: 1) Go to Elm Village 2) Click add application (add nothing) 3) Save new application 4) Click edit application (add nothing) 5) Try saving again and it will error out with the screenshot below. Note will this happened without entering any information, i was also able to reproduce it while entering preference geocoding info. Also note that I wasn't able to repro this on the target branch of this PR.
Screenshot 2024-01-30 at 4 04 12 PM

File: backend_new/src/controllers/map-layer.controller.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import { MapLayersService } from '../services/map-layers.service';
+import { MapLayerDto } from '../dtos/map-layers/map-layer.dto';
+import { mapTo } from '../utilities/mapTo';
+import { defaultValidationPipeOptions } from '../utilities/default-validation-pipe-options';
+import { OptionalAuthGuard } from '../guards/optional.guard';
+import { PermissionGuard } from '../guards/permission.guard';
+
+@Controller('/mapLayers')
+@ApiTags('mapLayers')
+@ApiBearerAuth()
```
Comment (YazeedLoonat) [inline backend_new/src/controllers/map-layer.controller.ts:20]: we don't need this in the new backend anymore

File: backend_new/src/controllers/map-layer.controller.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+@UsePipes(new ValidationPipe(defaultValidationPipeOptions))
+export class MapLayersController {
+  constructor(private readonly mapLayerService: MapLayersService) {}
+
+  @Get()
+  @ApiOperation({ summary: 'List map layers', operationId: 'list' })
+  async list(
+    @Query() queryParams: MapLayersQueryParams,
+  ): Promise<MapLayerDto[]> {
+    return mapTo(MapLayerDto, await this.mapLayerService.list(queryParams));
```
Comment (YazeedLoonat) [inline backend_new/src/controllers/map-layer.controller.ts:31]: can we move this mapTo to the service instead to bring it more inline with the other endpoints

File: backend_new/src/dtos/applications/application-multiselect-question-option.dto.ts [hunk 3]
```diff
@@ -64,6 +64,11 @@ export class ApplicationMultiselectQuestionOption {
   @ApiProperty()
   checked: boolean;
 
+  @Expose()
+  @IsString({ groups: [ValidationsGroupsEnum.default] })
+  @ApiPropertyOptional({ required: false })
```
Comment (YazeedLoonat) [inline backend_new/src/dtos/applications/application-multiselect-question-option.dto.ts:69]:
```suggestion
  @ApiPropertyOptional()
```

File: backend_new/src/dtos/map-layers/map-layer.dto.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { Expose } from 'class-transformer';
+import { IsString, IsUUID } from 'class-validator';
+import { ValidationsGroupsEnum } from '../../enums/shared/validation-groups-enum';
+
+export class MapLayerDto {
+  @Expose()
+  @IsString({ groups: [ValidationsGroupsEnum.default] })
+  @IsUUID(4, { groups: [ValidationsGroupsEnum.default] })
```
Comment (YazeedLoonat) [inline backend_new/src/dtos/map-layers/map-layer.dto.ts:8]: I think we need to add the @ApiProperty() notation for these

File: backend_new/src/dtos/map-layers/map-layers-query-params.dto.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { Expose } from 'class-transformer';
+import { ApiProperty } from '@nestjs/swagger';
+import { IsOptional, IsString } from 'class-validator';
+import { ValidationsGroupsEnum } from '../../enums/shared/validation-groups-enum';
+
+export class MapLayersQueryParams {
+  @Expose()
+  @ApiProperty({
```
Comment (YazeedLoonat) [inline backend_new/src/dtos/map-layers/map-layers-query-params.dto.ts:8]: we can switch this to the optional version and dropped the required:false

File: backend_new/src/dtos/multiselect-questions/multiselect-option.dto.ts [hunk 6]
```diff
@@ -72,4 +72,14 @@ export class MultiselectOption {
   @IsBoolean({ groups: [ValidationsGroupsEnum.default] })
   @ApiPropertyOptional()
   exclusive?: boolean;
+
+  @Expose()
+  @IsString({ groups: [ValidationsGroupsEnum.default] })
+  @ApiPropertyOptional({ required: false })
```
Comment (YazeedLoonat) [inline backend_new/src/dtos/multiselect-questions/multiselect-option.dto.ts:78]: same as an earlier comment for Optional we don't need required:false

**Review by YazeedLoonat — CHANGES_REQUESTED** [review]
Few things that I think we can tighten up as we bring stuff into prisma

**Comment by ColinBuyck** [comment]
Also seeing undefined and no address verification when submitting a paper application with preference address information

https://github.com/bloom-housing/bloom/assets/53269332/b9c62cc8-fd7b-4744-adc0-c570fc8ee9bd

**Comment by ludtkemorgan** [comment]
@ColinBuyck there was a bug where if a preference asked for address in more than one option it wasn't working which has been fixed. Also I realized the actual csv geocoding functionality was not ported over 😢. I have added that to this branch as well. Hopefully the functionality works correctly now

File: backend_new/test/integration/application.e2e-spec.ts [hunk 7]
```diff
@@ -486,7 +486,7 @@ describe('Application Controller Tests', () => {
       expect(mockApplicationConfirmation).toBeCalledTimes(1);
     });
 
-    it('should calculate geocoding on application', async () => {
+    it.only('should calculate geocoding on application', async () => {
```
Comment (YazeedLoonat) [inline backend_new/test/integration/application.e2e-spec.ts:489]:
```suggestion
    it('should calculate geocoding on application', async () => {
```

File: backend_new/src/controllers/map-layer.controller.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import { MapLayersQueryParams } from '../dtos/map-layers/map-layers-query-params.dto';
+import { MapLayersService } from '../services/map-layers.service';
+import { MapLayerDto } from '../dtos/map-layers/map-layer.dto';
+import { defaultValidationPipeOptions } from '../utilities/default-validation-pipe-options';
+import { OptionalAuthGuard } from '../guards/optional.guard';
+import { PermissionGuard } from '../guards/permission.guard';
+
+@Controller('/mapLayers')
+@ApiTags('mapLayers')
+@UseGuards(OptionalAuthGuard, PermissionGuard)
```
Comment (YazeedLoonat) [inline backend_new/src/controllers/map-layer.controller.ts:19]: question: do we need to add mapLayers to the permission csv?

File: shared-helpers/src/types/backend-swagger.ts [hunk 9]
```diff
@@ -1828,6 +1828,30 @@ export class AuthService {
...
+  /**
+   * List map layers
+   */
+  list(
+    params: {
+      /**  */
+      jurisdictionId?: string
+    } = {} as any,
+    options: IRequestOptions = {}
+  ): Promise<any> {
```
Comment (YazeedLoonat) [inline shared-helpers/src/types/backend-swagger.ts:1841]: I think this returning `any` indicates something is up with the controller type definition

or maybe the client generation needs to be re-run?

**Review by YazeedLoonat — CHANGES_REQUESTED** [review]
Looks great! just a few changes

File: backend_new/src/controllers/map-layer.controller.ts (hunk 8 above)
Comment (ludtkemorgan) [reply backend_new/src/controllers/map-layer.controller.ts:20]: Added!

File: shared-helpers/src/types/backend-swagger.ts (hunk 9 above)
Comment (ludtkemorgan) [reply shared-helpers/src/types/backend-swagger.ts]: I just needed to re-run!

**Review by YazeedLoonat — APPROVED** [review]
LGTM

**Pull request merged**